dlt/destinations/impl/lancedb/lancedb_client.py

```diff
@@ -0,0 +1,698 @@
+            )
+        return applied_update
+
+    def get_storage_table(self, table_name: str) -> Tuple[bool, TTableSchemaColumns]:
+        table_schema: TTableSchemaColumns = {}
+
+        try:
+            fq_table_name = self.make_qualified_table_name(table_name)
+            arrow_schema = self.db_client.open_table(fq_table_name).schema
+        except FileNotFoundError:
+            return False, table_schema
+        except Exception:
+            raise
+
+        for field_name, field_type in arrow_schema_to_dict(arrow_schema).items():
+            schema_c: TColumnSchema = {
+                "name": self.schema.naming.normalize_identifier(field_name),
+                **self._from_db_type(field_type, None, None),
+            }
+            table_schema[field_type] = schema_c
+        return True, table_schema
+
+    @lancedb_error
+    def add_table_field(self, table_name: str, field_schema: TArrowField) -> Optional[Table]:
+        """Add a field to the LanceDB table.
+
+        Since arrow tables are immutable, this is done via a staging mechanism.
+        The data is stored in-memory in a staging arrow table, evolved then stored
+        written over the old table.
+
+        Args:
+            table_name: The name of the table to create the field on.
+            field_schema: The field to create.
+        """
+        arrow_table = self.db_client.open_table(table_name).to_arrow()
+
+        # Create an array of null values for the new column.
+        null_array = pa.nulls(len(arrow_table), type=field_schema.type)
+
+        # Create staging Table with new column appended.
+        stage = arrow_table.append_column(field_schema, null_array)
+
+        try:
+            return self.db_client.create_table(table_name, stage, mode="overwrite")
+        except OSError:
+            # Field already present, skip.
+            return None
+        except Exception:
+            raise
+
+    def _execute_schema_update(self, only_tables: Iterable[str]) -> None:
+        for table_name in only_tables or self.schema.tables:
+            exists, existing_columns = self.get_storage_table(table_name)
+            new_columns = self.schema.get_new_table_columns(table_name, existing_columns)
+            embedding_fields: List[str] = get_columns_names_with_prop(
+                self.schema.get_table(table_name), VECTORIZE_HINT
+            )
+            logger.info(f"Found {len(new_columns)} updates for {table_name} in {self.schema.name}")
+            if len(new_columns) > 0:
+                if exists:
+                    for column in new_columns:
+                        field_schema: TArrowField = make_arrow_field_schema(
+                            column["name"], column, self.type_mapper, embedding_fields
+                        )
+                        fq_table_name = self.make_qualified_table_name(table_name)
+                        self.add_table_field(fq_table_name, field_schema)
```

Comment:
when adding multiple new fields this code will load the full table, add a new column, save it and repeat for every column. I think .add_table_field should take an array of new columns and add them at once. Or is there some other reason you solved it this way?

Reply:
Thanks for the spot!

There was no reason to do it this way, I've add supported for multiple fields in a single operation now with `add_table_fields`:

https://github.com/dlt-hub/dlt/blob/9b4c5196436f54c73a7fa61a8aa33bf5fe37f3b0/dlt/destinations/impl/lancedb/lancedb_client.py#L393-L422

Now the arrow table is mutated in memory, instead of read from the database n times over.